rust/processor/src/db/common/models/resources.rs

```diff
@@ -0,0 +1,212 @@
+    formatcp!("{COIN_ADDR}::fungible_asset::ConcurrentFungibleAssetBalance");
+
+pub const TYPE_OBJECT_CORE: &str = formatcp!("{COIN_ADDR}::object::ObjectCore");
+pub const TYPE_UNTRANSFERABLE: &str = formatcp!("{COIN_ADDR}::object::Untransferable");
+pub const TYPE_COLLECTION: &str = formatcp!("{TOKEN_V2_ADDR}::collection::Collection");
+pub const TYPE_CONCURRENT_SUPPLY: &str = formatcp!("{TOKEN_V2_ADDR}::collection::ConcurrentSupply");
+pub const TYPE_FIXED_SUPPLY: &str = formatcp!("{TOKEN_V2_ADDR}::collection::FixedSupply");
+pub const TYPE_UNLIMITED_SUPPLY: &str = formatcp!("{TOKEN_V2_ADDR}::collection::UnlimitedSupply");
+pub const TYPE_APOTS_COLLECTION: &str = formatcp!("{TOKEN_V2_ADDR}::aptos_token::AptosCollection");
+pub const TYPE_TOKEN_V2: &str = formatcp!("{TOKEN_V2_ADDR}::token::Token");
+pub const TYPE_TOKEN_IDENTIFIERS: &str = formatcp!("{TOKEN_V2_ADDR}::token::TokenIdentifiers");
+pub const TYPE_PROPERTY_MAP: &str = formatcp!("{TOKEN_V2_ADDR}::property_map::PropertyMap");
+
+pub trait Resource {
+    fn type_str() -> &'static str;
+}
+
+pub trait FromWriteResource<'a> {
+    fn from_write_resource(write_resource: &'a WriteResource) -> Result<Option<Self>>
+    where
+        Self: Sized;
+}
+
+impl<'a, T> FromWriteResource<'a> for T
+where
+    T: TryFrom<&'a WriteResource, Error = anyhow::Error> + Resource,
+{
+    fn from_write_resource(write_resource: &'a WriteResource) -> Result<Option<Self>> {
+        if MoveResource::get_outer_type_from_write_resource(write_resource) != Self::type_str() {
+            return Ok(None);
+        }
+        Ok(Some(write_resource.try_into()?))
+    }
+}
+
+pub enum V2FungibleAssetResource {
+    ConcurrentFungibleAssetBalance(ConcurrentFungibleAssetBalance),
+    ConcurrentFungibleAssetSupply(ConcurrentFungibleAssetSupply),
+    FungibleAssetMetadata(FungibleAssetMetadata),
+    FungibleAssetStore(FungibleAssetStore),
+    FungibleAssetSupply(FungibleAssetSupply),
+}
+
+impl Resource for ConcurrentFungibleAssetBalance {
+    fn type_str() -> &'static str {
+        TYPE_CONCURRENT_FUNGIBLE_ASSET_BALANCE
+    }
+}
+
+impl Resource for ConcurrentFungibleAssetSupply {
+    fn type_str() -> &'static str {
+        TYPE_CONCURRENT_FUNGIBLE_ASSET_SUPPLY
+    }
+}
+
+impl Resource for FungibleAssetMetadata {
+    fn type_str() -> &'static str {
+        TYPE_FUNGIBLE_ASSET_METADATA
+    }
+}
+
+impl Resource for FungibleAssetStore {
+    fn type_str() -> &'static str {
+        TYPE_FUNGIBLE_ASSET_STORE
+    }
+}
+
+impl Resource for FungibleAssetSupply {
+    fn type_str() -> &'static str {
+        TYPE_FUNGIBLE_ASSET_SUPPLY
+    }
+}
+
+impl V2FungibleAssetResource {
+    pub fn from_write_resource(write_resource: &WriteResource) -> Result<Option<Self>> {
+        let type_str = MoveResource::get_outer_type_from_write_resource(write_resource);
+        Ok(Some(match type_str.as_str() {
+            TYPE_CONCURRENT_FUNGIBLE_ASSET_BALANCE => {
```

Comment:
Use `ConcurrentFungibleAssetBalance::type_str()` here? Same for the other ones.

Reply:
It unfortunately doesn't compile, even it's compile time constant..

we can probably still write `x if x == ConcurrentFungibleAssetBalance::type_str() => {...}` but I feel it's ugly..